cmd/snap/cmd_run.go

```diff
@@ -530,6 +531,62 @@ func migrateXauthority(info *snap.Info) (string, error) {
 	return targetPath, nil
 }
 
+func activateXdgDocumentPortal(info *snap.Info) error {
+	// Don't do anything for snaps that don't plug the desktop interface
+	//
+	// NOTE: This check is imperfect because we don't really know
+	// if the interface is connected or not but this is an
+	// acceptable compromise for not having to communicate with
+	// snapd in snap run. In a typical desktop session the
+	// document portal can be in use by many applications, not
+	// just by snaps, so this is at most, pre-emptively using some
+	// extra memory.
+	plugsDesktop := false
+	for _, plug := range info.Plugs {
+		if plug.Interface == "desktop" {
+			plugsDesktop = true
+			break
+		}
+	}
+	if !plugsDesktop {
+		return nil
+	}
+
+	// If there is no session bus, our job is done.  We check this
+	// manually to avoid dbus.SessionBus() auto-launching a new
+	// bus.
+	busAddress := osGetenv("DBUS_SESSION_BUS_ADDRESS")
+	if len(busAddress) == 0 {
+		return nil
+	}
+
+	conn, err := dbus.SessionBus()
+	if err != nil {
+		return err
+	}
+
+	portal := conn.Object("org.freedesktop.portal.Documents",
+		"/org/freedesktop/portal/documents")
+	var mountPoint []byte
+	if err := portal.Call("org.freedesktop.portal.Documents.GetMountPoint", 0).Store(&mountPoint); err != nil {
+		// It is not considered an error if
+		// xdg-document-portal is not available on the system.
+		if dbusErr, ok := err.(dbus.Error); ok && dbusErr.Name == "org.freedesktop.DBus.Error.ServiceUnknown" {
+			return nil
+		}
+		return err
+	}
+
+	// Sanity check to make sure the document portal is exposed
+	// where we think it is.
+	actualMountPoint := strings.TrimRight(string(mountPoint), "\x00")
```

Comment:
Huh, I would've hoped go dbus had some helpers for this.

Reply:
The return type of the method is `ay` (array of bytes):

https://github.com/flatpak/xdg-desktop-portal/blob/master/data/org.freedesktop.portal.Documents.xml#L59

If it had been `s`, then go-dbus would have converted it to a Go string value.  Presumably this is done because paths on Linux are not guaranteed to be unicode.  It also seems that xdg-document-portal includes a nul byte at the end of the array for the benefit of C clients.